src/WebJobs.Extensions.DurableTask/Listener/TaskEntityShim.cs

```diff
@@ -388,60 +307,123 @@ private async Task ProcessOperationRequestAsync(RequestMessage request)
+                    };
+                    var guid = request.Id.ToString();
+                    this.context.SendEntityMessage(target, guid, responseMessage);
                 }
-                else
+            }
+
+            // send signal messages
+            foreach (var signal in outOfProcResult.Signals)
+            {
+                var request = new RequestMessage()
                 {
-                    this.Config.TraceHelper.FunctionCompleted(
-                        this.context.HubName,
-                        this.context.Name,
-                        this.context.InstanceId,
-                        request.Id.ToString(),
-                        request.Operation,
-                        this.Config.GetIntputOutputTrace(response.Result),
-                        continuedAsNew: true,
-                        functionType: FunctionType.Entity,
-                        isReplay: this.context.IsReplaying);
-                }
+                    ParentInstanceId = this.context.InstanceId,
+                    Id = Guid.NewGuid(),
+                    IsSignal = true,
+                    Operation = signal.Name,
+                    Input = signal.Input,
+                };
+                var target = new OrchestrationInstance()
+                {
+                    InstanceId = EntityId.GetSchedulerIdFromEntityId(signal.Target),
+                };
+                this.context.SendEntityMessage(target, "op", request);
             }
         }
 
-        public override void TraceAwait()
+        internal class OutOfProcResult
         {
-            // we only trace the awaits that are happening inside an operation,
-            // not the awaits that come from the scheduler loop
+            [JsonProperty("entityExists")]
+            public bool EntityExists { get; set; }
 
-            if (this.context.CurrentOperation != null)
+            [JsonProperty("entityState")]
+            public string EntityState { get; set; }
+
+            [JsonProperty("results")]
+            public List<OperationResult> Results { get; set; }
+
+            [JsonProperty("signals")]
+            public List<Signal> Signals { get; set; }
+
+            public struct OperationResult
             {
-                this.Config.TraceHelper.FunctionAwaited(
-                    this.context.HubName,
-                    this.context.Name,
-                    this.context.FunctionType,
-                    this.context.InstanceId,
-                    this.context.CurrentOperation?.Id.ToString() ?? string.Empty,
-                    this.context.CurrentOperation?.Operation ?? string.Empty,
-                    this.context.IsReplaying);
+                [JsonProperty("result")]
+                public string Result { get; set; }
+
+                [JsonProperty("isError")]
+                public bool IsError { get; set; }
+
+                [JsonProperty("duration")]
+                public double Duration { get; set; }
```

Comment:
What is the unit for `Duration`? Can we add a comment, or use the name of the field to make this clear?

Reply:
Yes to both. I will add comments to all the fields.